File: web/src/domain/requestLimits.ts
// P4-19(issue 110、ADR-0208): calc の候補・観測の件数上限。api/openapi.yaml の BulkCalcRequest.itemVariants /
// ReverseRequest.itemCandidates / ReverseRequest.observations の maxItems と同じ値を Web でも持つ。
//
// なぜ Web にも持つのか: 上限を超えた配列を送ると API は 400 invalid_input、engine(WASM)は同じ上限の
// sentinel で失敗するので、画面が黙って壊れる。上限に当たることが分かるのは入力を組み立てる画面側だけなので、
// ここで決定的に絞り込み、絞り込んだことを利用者に見せる(黙って切り捨てない。DECISIONS.md 2026-09-23)。
//
// 値の正は api/openapi.yaml(実行時に YAML は読めないので、この定数はその写し)。ずれたら
// requestLimits.test.ts が openapi.yaml を読んで失敗する(コーディング規約 §2「同期を検査するテストを置く」)。

/** 一括計算に渡せる持ち物の通り数(`BulkCalcRequest.itemVariants` の maxItems)。null(持ち物なし)も1通りと数える。 */
export const MAX_ITEM_VARIANTS = 64;

/** 逆算で探索できる持ち物の通り数(`ReverseRequest.itemCandidates` の maxItems)。null(持ち物なし)も1通りと数える。 */
export const MAX_ITEM_CANDIDATES = 64;

/** 逆算に渡せる観測の件数(`ReverseRequest.observations` の maxItems)。 */
export const MAX_OBSERVATIONS = 16;

/** 上限で絞り込んだ結果。truncated は「落とした要素があるか」(ちょうど上限のときは false)。 */
export interface LimitedValues<T> {
  readonly values: readonly T[];
  readonly truncated: boolean;
}

/**
 * 上限までを先頭から取る(マスタの順序をそのまま使う既存の規約。ADR-0300 §6・§7)。
 * 並べ替え・間引きはせず、末尾から落とすだけなので、同じ入力からは常に同じ結果になる。
 */
export function limitToMax<T>(values: readonly T[], max: number): LimitedValues<T> {
  return values.length <= max
    ? { values, truncated: false }
    : { values: values.slice(0, max), truncated: true };
}

File: ios/PokeCalcKit/Sources/PokeCalcCore/RequestLimits.swift
/// 計算・逆算の要求に含められる件数の上限(issue #110。ADR-0501「issue #110 の受け入れ条件(iOS 側)」
/// 2章・8章)。
///
/// 正は `api/openapi.yaml` の `maxItems`(`ReverseRequest.observations` / `ReverseRequest.itemCandidates` /
/// `BulkCalcRequest.itemVariants`)。この enum はその写しで、ViewModel・View はここだけを参照し、
/// 件数を直書きしない(coding-rules §2)。契約と写しのずれは `ios/scripts/check-request-limits.sh`
/// (`make ios-test` の一部)が検出する。
public enum RequestLimits {
    /// `ReverseRequest.observations.maxItems`。
    public static let maxObservations = 16
    /// `ReverseRequest.itemCandidates.maxItems`。
    public static let maxItemCandidates = 64
    /// `BulkCalcRequest.itemVariants.maxItems`。
    public static let maxItemVariants = 64

    /// 送る配列の先頭に入る null(持ち物なし)の1件を除いた、トグルで選べる ID の数
    /// (ADR「issue #110」3章)。
    public static let maxSelectableItemCandidates = maxItemCandidates - 1
    /// 同上(計算画面の比較トグル)。
    public static let maxSelectableItemVariants = maxItemVariants - 1

    /// `getMovesByIds` の `ids`(クエリパラメータ)の `maxItems`。1回の呼び出しに渡せる技 ID の上限
    /// (ADR-0501「getMovesByIds による構築編集の技の一括解決」1章)。`PokeCalcService.moves(ids:)` は
    /// これを超える集合をこの件数ずつに分割して複数回呼ぶ。`ReverseRequest.observations` 等と違い、
    /// `components.schemas` のプロパティではなく `paths./api/pokedex/moves/batch.get` のクエリパラメータ
    /// なので、`check-request-limits.sh` の照合はスキーマの `maxItems` とは別の経路で行う。
    public static let maxMoveBatchIds = 64
}

/// 件数の上限に達したことを画面に出す文言(`MasterSearchLabels` と同じ理由でコードに1か所持つ)。
/// 件数は `RequestLimits` から埋め込む(数字を文字列に直書きしない)。
public enum RequestLimitLabels {
    public static let observationsReachedLimit =
        "観測は最大\(RequestLimits.maxObservations)件までです"
    public static let itemCandidatesReachedLimit =
        "持ち物の候補は「持ち物なし」を含めて最大\(RequestLimits.maxItemCandidates)件までです"
    public static let itemVariantsReachedLimit =
        "比較する持ち物は「持ち物なし」を含めて最大\(RequestLimits.maxItemVariants)件までです"
}
